scritchPencilRaster: Bring in most of the missing Graphics methods

AShiningRay opened the pull request:
This commit implements the following methods in Software:
- drawArc
- drawRoundRect
- fillArc
- fillRoundRect
- fillTriangle
- fillPolygon (will be used by DirectGraphics and DoJa)

Thus making SquirrelJME's graphics subsystem a lot more robust and removing many cases where it would previously fail with a TODO call. Fixes games such as Sonic 2 (won't boot without the staged DirectGraphics work), Landers, Bounce Tales (also needs DirectGraphics), among others.

Furthermore, new corePrim_ methods were added for those, and fill/drawRect as well, in order to facilitate Hardware implementations for them when the time comes.

Also, native calls to drawTriangle were removed due to them not being actually implemented natively, and they effectively being replaced by three calls to drawLine, which shouldn't cause much overhead...

```
You grant Stephanie Gawroriski an irrevocable license that:

 1. That you own the contributing work.
 2. Grants a patent license, as per the GNU GPLv3.
 3. Granting Stephanie Gawroriski permission to redistribute, sell, lease,
    modify, transform, translate, and relicense the specified works. This
    is to simplify the licensing of the project and permit it to be
    consistent. Your contribution may be commercially licensed to other
    parties supporting the project through this means.
 4. If employed by a company, you have a right by that company to provide
    contributions to this project.
 5. Have pledged to follow the Code of Conduct.
 6. Have read and understand the Ettiquite.
```

 * I accept the contributor agreement.

 * I accept that SquirrelJME uses Fossil for its source control
   management and I do understand that my pull request will be merged
   via patch in Fossil instead of via GitHub.

AShiningRay committed:
scritchPencilRaster: Bring in most of the missing Graphics methods

This commit implements the following methods in Software:
- drawArc
- drawRoundRect
- fillArc
- fillRoundRect
- fillTriangle
- fillPolygon (will be used by DirectGraphics and DoJa)

Thus making SquirrelJME's graphics subsystem a lot more robust and
removing many cases where it would previously fail with a TODO call. Fixes
games such as Sonic 2 (won't boot without the staged DirectGraphics work),
Landers, Bounce Tales (also needs DirectGraphics), among others.

Furthermore, new corePrim_ methods were added for those, and fill/drawRect
as well, in order to facilitate Hardware implementations for them when
the time comes.

Also, native calls to drawTriangle were removed due to them not being
actually implemented natively, and they effectively being replaced
by three calls to drawLine, which shouldn't cause much overhead...

XerTheSquirrel commented on nanocoat/include/lib/scritchui/scritchuiPencil.h:
```diff
@@ -466,29 +510,6 @@ typedef sjme_errorCode (*sjme_scritchui_pencilDrawSubstringFunc)(
-	sjme_attrInValue sjme_jint y2,
-	sjme_attrInValue sjme_jint x3,
-	sjme_attrInValue sjme_jint y3);
-
```
This should not have been deleted.

XerTheSquirrel commented:
> Also, native calls to drawTriangle were removed due to them not being actually implemented natively, and they effectively being replaced by three calls to drawLine, which shouldn't cause much overhead...

`drawTriangle` must remain, they might not be native however they are post buffer lock, additionally there is reduced overhead having everything stay in ScritchUI rather than Java.

XerTheSquirrel commented on nanocoat/include/lib/scritchui/scritchuiPencil.h:
```diff
@@ -1114,25 +1204,34 @@ typedef struct sjme_scritchui_pencilFunctions
 	/** @c DrawSubstring . */
 	SJME_SCRITCHUI_QUICK_PENCIL(DrawSubstring, drawSubstring);
 	
-	/** @c DrawTriangle . */
-	SJME_SCRITCHUI_QUICK_PENCIL(DrawTriangle, drawTriangle);
-	
```
This should not have been deleted.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencil.c:
```diff
@@ -107,16 +107,20 @@ static const sjme_scritchui_pencilFunctions sjme_scritchpen_core_functions =
 	sjme_sm(.drawRect, sjme_scritchpen_core_drawRect),
+	sjme_sm(.drawRoundRect, sjme_scritchpen_core_drawRoundRect),
 	sjme_sm(.drawSubstring, sjme_scritchpen_core_drawSubstring),
-	sjme_sm(.drawTriangle, sjme_scritchpen_core_drawTriangle),
```
This should not have been deleted.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencil.c:
```diff
@@ -136,16 +140,20 @@ static const sjme_scritchui_pencilFunctions
 	sjme_sm(.drawRect, sjme_scritchpen_coreSerial_drawRect),
+	sjme_sm(.drawRoundRect, sjme_scritchpen_coreSerial_drawRoundRect),
 	sjme_sm(.drawSubstring, sjme_scritchpen_coreSerial_drawSubstring),
-	sjme_sm(.drawTriangle, sjme_scritchpen_coreSerial_drawTriangle),
```
This should not have been deleted.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchSerialize.c:
```diff
@@ -1708,66 +1747,45 @@ sjme_errorCode sjme_scritchpen_coreSerial_copyArea(
 	SJME_SDX_PASS(x);
 	SJME_SDX_PASS(y);
 	SJME_SDX_PASS(w);
-	SJME_SDX_PASS(h);
-	
-	/* Invoke and wait. */
-	SJME_SDX_WAIT;
-}
-
-sjme_errorCode sjme_scritchpen_coreSerial_drawTriangle(
-	sjme_attrInNotNull sjme_scritchui_pencil g,
-	sjme_attrInValue sjme_jint x1,
-	sjme_attrInValue sjme_jint y1,
-	sjme_attrInValue sjme_jint x2,
-	sjme_attrInValue sjme_jint y2,
-	sjme_attrInValue sjme_jint x3,
-	sjme_attrInValue sjme_jint y3)
-{
-	SJME_SDP_CHUNK(drawTriangle,
-		SJME_SCRITCHUI_SERIAL_PEN_DRAW_TRIANGLE,
-		(g, x1, y1, x2, y2, x3, y3));
-		
-	SJME_SDX_PASS(g);
-	SJME_SDX_PASS(x1);
-	SJME_SDX_PASS(y1);
-	SJME_SDX_PASS(x2);
-	SJME_SDX_PASS(y2);
-	SJME_SDX_PASS(x3);
-	SJME_SDX_PASS(y3);
```
This should not have been deleted.

XerTheSquirrel commented on emulators/springcoat-vm/src/main/java/cc/squirreljme/vm/springcoat/MLEPencil.java:
```diff
@@ -74,6 +74,46 @@ public Object handle(SpringThreadWorker __thread, Object... __args)
+	 * int, int)}.
+	 */
+	HARDWARE_Draw_ARC("hardwareDrawArc:" +
+		"(Lcc/squirreljme/jvm/mle/brackets/PencilBracket;IIIIII)V")
```
There is the newer and more precise `MLEDispatcher.methodKey()` for declaring method signatures in SpringCoat here.

AShiningRay replied:
I think it'd be better to keep them defined as they are right now for consistency, since it is how all the other methods are defined in the file as well... then have another separate commit just moving everyone to MLEDispatcher.methodKey() later on.

XerTheSquirrel commented on emulators/springcoat-vm/src/main/java/cc/squirreljme/vm/springcoat/MLEPencil.java:
```diff
@@ -287,6 +327,85 @@ public Object handle(SpringThreadWorker __thread, Object... __args)
+	 * {@link PencilShelf#hardwareFillArc(PencilBracket, int, int, int, int,
+	 * int, int)}.
+	 */
+	HARDWARE_FILL_ARC("hardwareFillArc:" +
+		"(Lcc/squirreljme/jvm/mle/brackets/PencilBracket;IIIIII)V")
```
There is the newer and more precise `MLEDispatcher.methodKey()` for declaring method signatures in SpringCoat here.

XerTheSquirrel commented on emulators/springcoat-vm/src/main/java/cc/squirreljme/vm/springcoat/MLEPencil.java:
```diff
@@ -287,6 +327,85 @@ public Object handle(SpringThreadWorker __thread, Object... __args)
+	 * {@link PencilShelf#hardwareFillPolygon(PencilBracket, int[], int, int[],
+	 * int, int)}.
+	 */
+	HARDWARE_FILL_POLYGON("hardwareFillPolygon:" +
+		"(Lcc/squirreljme/jvm/mle/brackets/PencilBracket;[II[III)V")
```
There is the newer and more precise `MLEDispatcher.methodKey()` for declaring method signatures in SpringCoat here.

XerTheSquirrel commented on emulators/springcoat-vm/src/main/java/cc/squirreljme/vm/springcoat/MLEPencil.java:
```diff
@@ -321,6 +440,46 @@ public Object handle(SpringThreadWorker __thread, Object... __args)
+	 * {@link PencilShelf#hardwareFillRoundRect(PencilBracket, int, int, int,
+	 * int, int, int)}.
+	 */
+	HARDWARE_FILL_ROUND_RECT("hardwareFillRoundRect:" +
+		"(Lcc/squirreljme/jvm/mle/brackets/PencilBracket;IIIIII)V")
```
There is the newer and more precise `MLEDispatcher.methodKey()` for declaring method signatures in SpringCoat here.

XerTheSquirrel commented on nanocoat/include/lib/scritchui/core/coreRaster.h:
```diff
@@ -311,14 +369,14 @@ sjme_errorCode sjme_scritchpen_core_drawRect(
+	sjme_attrInPositive sjme_jint w,
+	sjme_attrInPositive sjme_jint h,
+	sjme_attrInPositive sjme_jint arcWidth,
+	sjme_attrInPositive sjme_jint arcHeight);
```
Please keep `drawTriangle()`.

XerTheSquirrel commented on nanocoat/include/lib/scritchui/core/coreSerial.h:
```diff
@@ -227,36 +227,48 @@ typedef enum sjme_scritchui_serialType
-	/** @c drawRect . */
-	SJME_SCRITCHUI_SERIAL_PEN_DRAW_RECT,
-
-	/** @c drawTriangle . */
-	SJME_SCRITCHUI_SERIAL_PEN_DRAW_TRIANGLE,
```
Please keep `drawTriangle()`.

XerTheSquirrel commented on nanocoat/include/lib/scritchui/core/coreSerial.h:
```diff
@@ -603,27 +615,14 @@ SDP_STRUCT_DEF(copyArea,
 	SDX_VAR(sjme_jint, w);
-	SDX_VAR(sjme_jint, h););
-
-SDP_STRUCT_DEF(drawTriangle,
-	SDX_VAR(sjme_scritchui_pencil, g);
-	SDX_VAR(sjme_jint, x1);
-	SDX_VAR(sjme_jint, y1);
-	SDX_VAR(sjme_jint, x2);
-	SDX_VAR(sjme_jint, y2);
-	SDX_VAR(sjme_jint, x3);
-	SDX_VAR(sjme_jint, y3););
```
Please keep `drawTriangle()`.

XerTheSquirrel commented on nanocoat/include/lib/scritchui/core/coreSerial.h:
```diff
@@ -831,16 +879,20 @@ typedef union sjme_scritchui_serialDataUnion
+	SJME_SCRITCHUI_SDP_DEF(drawArc);
 	SJME_SCRITCHUI_SDP_DEF(drawHoriz);
 	SJME_SCRITCHUI_SDP_DEF(drawRect);
-	SJME_SCRITCHUI_SDP_DEF(drawTriangle);
```
Please keep `drawTriangle()`.

XerTheSquirrel commented on nanocoat/include/lib/scritchui/core/coreSerial.h:
```diff
@@ -1210,27 +1262,14 @@ sjme_errorCode sjme_scritchpen_coreSerial_copyArea(
 	sjme_attrInPositive sjme_jint w,
-	sjme_attrInPositive sjme_jint h);
-
-sjme_errorCode sjme_scritchpen_coreSerial_drawTriangle(
-	sjme_attrInNotNull sjme_scritchui_pencil g,
-	sjme_attrInValue sjme_jint x1,
-	sjme_attrInValue sjme_jint y1,
-	sjme_attrInValue sjme_jint x2,
-	sjme_attrInValue sjme_jint y2,
-	sjme_attrInValue sjme_jint x3,
-	sjme_attrInValue sjme_jint y3);
```
Please keep `drawTriangle()`.

XerTheSquirrel commented on nanocoat/include/lib/scritchui/scritchuiPencil.h:
```diff
@@ -534,6 +555,50 @@ typedef sjme_errorCode (*sjme_scritchui_pencilDrawXRGB32RegionFunc)(
+		sjme_attrInNotNull sjme_jint* yPoints,
+		sjme_attrInPositive sjme_jint yOffset,
+		sjme_attrInPositive sjme_jint nPoints);
+
```
Indent left.

For `xPoints` and `yPoints`, they should be `sjme_attrInNotNullBuf(nPoints)`.

The `@since` is not the current date.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	sjme_errorCode error;
+	sjme_jint yz, yze;
+	float centerX, centerY, radiusX, radiusY, startAngleRad, endAngleRad,
+		angle;
```
Split `angle` into its own declaration of `float` since it is too long to fit.

`float` should be `sjme_jfloat`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	sjme_attrInValue sjme_jint startAngle,
+	sjme_attrInValue sjme_jint arcAngle)
+{
+	sjme_errorCode error;
```
Guard function with


```c
#if defined(SJME_CONFIG_HAS_FLOAT_HARD)
	...
#else
	sjme_todo("Impl?");
	return sjme_error_notImplemented(0);
#endif
```

AShiningRay replied:
Since this will already guard all of the float usage in the function, and is the same guard as for sjme_jfloat itself i'd rather not use sjme_jfloat explicitly since it is a union, which ended up making parts of fillArc and drawArc quite cumbersome to read due to all the manipulation.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	sjme_scritchui_pencilDrawHorizFunc drawHoriz;
+	sjme_jint yz, yze, steps, innerX, innerY;
+	sjme_jboolean hasAlpha = 0;
+	float centerX, centerY, radiusX, radiusY, startAngleRad, endAngleRad,
```
Should be `sjme_jfloat`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	sjme_jint yz, yze, steps, innerX, innerY;
+	sjme_jboolean hasAlpha = 0;
+	float centerX, centerY, radiusX, radiusY, startAngleRad, endAngleRad,
+		maxRad, angle;
```
Overflow should be on next line as `sjme_jfloat` rather than being indented.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	sjme_errorCode error;
+	sjme_scritchui_pencilDrawHorizFunc drawHoriz;
+	sjme_jint yz, yze, steps, innerX, innerY;
+	sjme_jboolean hasAlpha = 0;
```
Initializer should be `SJME_JNI_FALSE`, it should also be later after everything as not all compilers support mixed initialization.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	sjme_jboolean hasAlpha = 0;
+	float centerX, centerY, radiusX, radiusY, startAngleRad, endAngleRad,
+		maxRad, angle;
+	sjme_jboolean *filledPixels = NULL;
```
Initializer should be later after everything as not all compilers support mixed initialization.

Additionally, star on `sjme_jboolean`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	sjme_attrInValue sjme_jint startAngle,
+	sjme_attrInValue sjme_jint arcAngle)
+{
+	sjme_errorCode error;
```
Guard function with


```c
#if defined(SJME_CONFIG_HAS_FLOAT_HARD)
	...
#else
	sjme_todo("Impl?");
	return sjme_error_notImplemented(0);
#endif
```

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	return error;
+	
+fail_any:
+	
```
`filledPixels` is never freed here if allocated.

AShiningRay replied:
The only goto that will happen before deallocating filledPixels in here is in the allocation block itself, so it should be fine as it wouldn't have been allocated anyway?

```
if(hasAlpha)
	{
		/* We add 1 to the allocation size just so ceil() isn't needed */
		allocSize = (w * h) + 1;
		filledPixels = (sjme_jboolean *) sjme_alloca(allocSize);
		if (filledPixels == NULL)
		{
			error = sjme_error_outOfMemory(NULL, allocSize);
			goto fail_any;
		}
		memset(filledPixels, 0, allocSize);
	}
```

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+			 * Only draw if opaque, or if the alpha buffer is not yet filled
+			 * for the current position.
+			 */
+			if(hasAlpha ? !*(filledPixels + ((innerY-y) * w + innerX-x)) : 1)
```
Extra parenthesis inner ternary, for readability and so one does not have to double check the operator precedence chart.

XerTheSquirrel replied:
`filledPixels[]` instead of `*filledPixels`, as it is more concise and readable.

XerTheSquirrel replied:
Something like `(hasAlpha ? (filledPixels[filledZ >> 3] & (1 << (filledZ & 7))) : 1)`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+		float angle = startAngleRad + (i * endAngleRad / steps);
+		
+		int innerX = round((centerX) + radiusX * cos(angle));
+		int innerY = round((centerY) + radiusY * sin(angle));
```
These variable declarations must be moved to the start of the function.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	
+	for (int i = 0; i < steps; i++) 
+	{
+		float angle = startAngleRad + (i * endAngleRad / steps);
```
Parenthesis `i * endAngleRad`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	radiusX = w / 2.0f;
+	radiusY = h / 2.0f;
+	startAngleRad = startAngle * 0.017453292f;
+	endAngleRad = ((startAngle + arcAngle) * 0.017453292f) 
```
`#define SJME_ANGLE_RAD 0.017453292f` for magic number `0.017453292f`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	startAngleRad = startAngle * 0.017453292f;
+	endAngleRad = ((startAngle + arcAngle) * 0.017453292f) 
+	- startAngleRad;
+	steps = fabs(arcAngle * ((w + h) / 2.0f) / 50.0f);
```
Parenthesis `arcAngle * ((w + h) / 2.0f)`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	radiusY = h / 2.0f;
+	startAngleRad = startAngle * 0.017453292f;
+	endAngleRad = ((startAngle + arcAngle) * 0.017453292f) 
+	- startAngleRad;
```
Indent continuation, minus operator should be on the previous line.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	
+	if (g == NULL)
+		return SJME_ERROR_NULL_ARGUMENTS;
+	
```
Parameters `w` and `h` are `sjme_attrInPositive` however they are not checked.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+		
+		/* Make sure we're not drawing out of bounds. */
+		if(innerX < clipLine->s.x || innerX >= clipLine->e.x
+			|| innerY < clipLine->s.y || innerY >= clipLine->e.y)
```
`||` previous line at end.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	steps = fabs(arcAngle * ((w + h) / 2.0f) / 50.0f);
+	innerX, innerY;
+
+	firstFillX = lastFillX = firstFillY = lastFillY = -1;
```
```c
lastFillX = -1;
lastFillY = -1;
firstFillX = round((centerX) + radiusX * cos(startAngleRad));
firstFillY = round((centerY) + radiusY * sin(startAngleRad)); 
```

AShiningRay replied:
This one is for circles only, where the last pixel ends up at the same point of the first one... Wouldn't this cause issues because:
```
if((lastFillX == innerX && lastFillY == innerY) || 
	(firstFillX == innerX && firstFillY == innerY)) 
	continue;
```

Will evaluate to true for the first pixel drawn as well, and not only the last? One way i see this working is by drawing the first pixel out of that loop:
```
      lastFillX = -1;
	lastFillY = -1;
	firstFillX = round((centerX) + radiusX * cos(startAngleRad));
	firstFillY = round((centerY) + radiusY * sin(startAngleRad)); 

	/* Make sure we're not drawing out of bounds. */
	if(firstFillX >= clipLine->s.x || firstFillX < clipLine->e.x ||
		firstFillY >= clipLine->s.y || firstFillY < clipLine->e.y)
	{
		/* If style is DOTTED, rendering will paint and skip pixels 1 by 1. */
		if (dot)
			error |= drawPixel(g, firstFillX, firstFillY);
		dot ^= dotFlip;
	}
	
	for (i = 0; i < steps; i++) 
	{
	      ...
```

And then removing the `if i == 0` check accordingly since the first pixel will already be drawn anyway (if not out of bounds, in which case the last pixel won't be either since it's a circle)

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+		{ 
+			firstFillY = innerY; 
+			firstFillX = innerX; 
+		}
```
Delete this in favor of <https://github.com/SquirrelJME/SquirrelJME/pull/252/files#r2574339082>.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	sjme_jint yz, yze;
+	float centerX, centerY, radiusX, radiusY, startAngleRad, endAngleRad,
+		angle;
+	int steps, innerX, innerY, firstFillX, lastFillX, firstFillY, lastFillY;
```
`sjme_jint`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+
+	if (g == NULL)
+		return SJME_ERROR_NULL_ARGUMENTS;
+		
```
Parameters `w` and `h` are not checked here.

AShiningRay replied:
Those checks will be made in the core_ variants of those, since corePrim ones are called by other functions as well, like fill/drawRoundRect, which themselves also check for w and h arguments.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+				sizeof(sjme_jboolean));
+			goto fail_any;
+		}
+		memset(filledPixels, 0, (w * h) * sizeof(sjme_jboolean));
```
`(w * h) * sizeof(sjme_jboolean)` can be stored into a variable `allocLen` or similar, reducing the need to have it copied everywhere.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+			if(hasAlpha ? !*(filledPixels + ((innerY-y) * w + innerX-x)) : 1)
+			{
+				if(hasAlpha)
+					*(filledPixels + ((innerY-y) * w + innerX-x)) = 1;
```
`filledPixels[]` is more readable and concise than `*filledPixels`

XerTheSquirrel replied:
Something like `filledPixels[filledZ >> 3] |= (1 << (filledZ & 7))`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+				|| innerY < clipLine->s.y || innerY >= clipLine->e.y 
+				|| innerX-x < 0 || innerY-y < 0)
+				continue;
+
```
Declare and set `filledZ = ((innerY-y) * w + innerX-x)`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	 *
+	 * TODO: This can be optimized to take 1/8 of the usual memory size by
+	 * manipulating each bit of the boolean array below, as a _Bool variable
+	 * is often at least one whole byte in size.
```
With `sjme_jint filledZ` and something like `(hasAlpha ? (filledPixels[filledZ >> 3] & (1 << (filledZ & 7))) : 1)` and `filledPixels[filledZ >> 3] |= (1 << (filledZ & 7))`, this can be reduced by 8 right now and would be preferrable.

AShiningRay replied:
I changed the attributions to `(1 << (7 - filledZ & 7))` instead, as `(1 << (filledZ & 7))` probably wouldn't work as intended because pixels are arranged left to right. The end result is thus:

```
      if(hasAlpha ? (filledPixels[filledZ >> 3] &
		(1 << (7 - filledZ & 7))) : SJME_JNI_TRUE)
	{
		if(hasAlpha)
			filledPixels[filledZ >> 3] |= (1 << (7 - filledZ & 7));
		error |= drawPixel(g, innerX, innerY);
	}
```

Works fine.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	radiusY = h / 2.0f;
+	startAngleRad = startAngle * 0.017453292f;
+	endAngleRad = ((startAngle + arcAngle) * 0.017453292f) 
+	- startAngleRad;
```
`#define` the magic number `0.017453292f`.

`-` should be on the previous line and `startAngleRad` should be a continuation indent.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	 * At least it has the upside of generating far more stable arcs that fully
+	 * match their outline compared to Java AWT's algorithm.
+	 * 
+	 * TODO: Optimize this later
```
Remove this `TODO`, there is always an implied later optimization once something works.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	int ymin = 2147483647, ymax = -2147483648, intersectionCount = 0, xStart,
+	xEnd, temp, dy, j;
+	int intersections[nPoints];
+	long ix;
```
`sjme_jint` for `int` and `sjme_jlongNative` for `long` respectfully.

The initialization should be moved out as initializer mixing is not possible with all compilers.

Additionally, there is `INT32_MAX` and `INT32_MIN` for those magic numbers.

AShiningRay replied:
The jlongNative variable was made into a sjme_jint as well. The long was mostly an "additional safety measure" i put in place when testing edge cases in FreeJ2ME-Plus, the likelihood of the arithmetic here going out of range of a standard int are microscopic, if not nonexistent.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+
+	if (g == NULL)
+		return SJME_ERROR_NULL_ARGUMENTS;
+
```
`xOffset`, `yOffset`, and `nPoints` must be checked for bounds and overflow accordingly.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	sjme_attrInNotNull sjme_scritchui_pencil g,
+	sjme_attrInValue sjme_jint xPoints[],
+	sjme_attrInPositive sjme_jint xOffset,
+	sjme_attrInValue sjme_jint yPoints[],
```
Signature should use pointers and not arrays, and to also match previous declarations.

```c
		sjme_attrInNotNull sjme_jint* xPoints,
		sjme_attrInNotNull sjme_jint* yPoints,
```

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	for (int i = 0; i < nPoints; i++) 
+	{
+		if (yPoints[i+yOffset] < ymin) { ymin = yPoints[i+yOffset]; }
+		if (yPoints[i+yOffset] > ymax) { ymax = yPoints[i+yOffset]; }
```
Does not match code style, needs more whitespace and additionally the variable sets do not need brackets.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	{
+		angle = startAngleRad + (i * endAngleRad / steps);
+
+		for(int j = 0; j < maxRad; j++) 
```
`int j` is not valid here, it needs to be declared at the start of function as `sjme_jint j`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	steps = fabs(arcAngle * ((w + h) / 2.0f) / 50.0f);
+	innerX, innerY;
+
+	for(int i = 0; i < steps; i++) 
```
`int i` is not valid here, it needs to be declared at the start of function as `sjme_jint i`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+
+	firstFillX = lastFillX = firstFillY = lastFillY = -1;
+	
+	for (int i = 0; i < steps; i++) 
```
`int i` is not valid here, it needs to be declared at the start of function as `sjme_jint i`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	 * intersections by increasing X coordinate, then fill from top to bottom.
+	 */
+
+	for (int i = 0; i < nPoints; i++) 
```
`int i` is not valid here, it needs to be declared at the start of function as `sjme_jint i`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	if(ymax + g->state.translateReal.y >= clipLine->e.y) 
+		ymax = clipLine->e.y - g->state.translateReal.y;
+
+	for (int y = ymin; y < ymax; y++)
```
`int y` is not valid here, it needs to be declared at the start of function as `sjme_jint y`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+	for (int y = ymin; y < ymax; y++)
+	{
+		intersectionCount = 0;
+		for (int i = 0; i < nPoints; i++)
```
`int i` is not valid here, it needs to be declared at the start of function as `sjme_jint i`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+				dy = yPoints[j + yOffset] - yPoints[i + yOffset];
+				if (dy != 0)
+				{
+					ix = (long) xPoints[i + xOffset] * dy + (y - yPoints[i +
```
Cast to `(sjme_jlongNative)`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+				if (dy != 0)
+				{
+					ix = (long) xPoints[i + xOffset] * dy + (y - yPoints[i +
+							yOffset]) * (long) (xPoints[j + xOffset] -
```
Cast to `(sjme_jlongNative)`

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+							yOffset]) * (long) (xPoints[j + xOffset] -
+							xPoints[i + xOffset]);
+					ix /= dy;
+					intersections[intersectionCount++] = (int) ix;
```
Cast to `(sjme_jint)`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+			if (i + 1 < intersectionCount)
+			{
+				xStart = fmax(intersections[i], clipLine->s.x);
+				xEnd = fmin(intersections[i + 1], clipLine->e.x);
```
Needless conversion to/from `float`, this would be better using `min/max`, if missing those can be added in `util.h`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+
+		for (int i = 0; i < intersectionCount - 1; i++)
+		{
+			for (int j = 0; j < intersectionCount - 1 - i; j++)
```
`int i, j` are not valid here, they need to be declared at the top of the function.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+			}
+		}
+
+		for (int i = 0; i < intersectionCount; i += 2)
```
`int i` is not valid here, it needs to be declared at the top of the function.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -17,7 +17,399 @@
+				if(xEnd > xStart)
+					if (sjme_error_is(error = drawHoriz(g, xStart, y,
+						xEnd-xStart)))
+						break;
```
`error |= `, this reduces overall branching.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -60,12 +445,86 @@ sjme_errorCode sjme_scritchpen_core_drawRect(
+	drawHoriz = g->prim.drawHoriz;
+	for (yz = y, yze = y + h; yz < yze; yz++)
+		if (sjme_error_is(error = drawHoriz(g, x, yz, w)))
+			break;
```
`error |=` to reduce overall branching.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -60,12 +445,86 @@ sjme_errorCode sjme_scritchpen_core_drawRect(
+			break;
+	
+	/* Failed? */
+	if (sjme_error_is(error))
+		goto fail_any;
+		
+	/* Success? */
+	return error;
+	
+fail_any:
+	
+	return sjme_error_default(error);
```
```c
	/* Failed? */
	if (sjme_error_is(error))
		return sjme_error_default(error);

	/* Success! */
	return SJME_ERROR_NONE;
```

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -60,12 +445,86 @@ sjme_errorCode sjme_scritchpen_core_drawRect(
+	error |= drawArc(g, x, y, w, h, startAngle, arcAngle);
+	
+	/* Failed? */
+	if (sjme_error_is(error))
```
This can be combined into:

```c
if (sjme_error_is(error = g->prim.drawArc(g, x, y, w, h, startAngle, arcAngle)))
```
```

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -74,47 +533,233 @@ sjme_errorCode sjme_scritchpen_core_drawRect(
 	return sjme_error_default(error);
 }
 
-sjme_errorCode sjme_scritchpen_core_drawTriangle(
```
Please bring back `drawTriangle()`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -74,47 +533,233 @@ sjme_errorCode sjme_scritchpen_core_drawRect(
+	error |= drawRect(g, x, y, w, h);
+	
+	/* Failed? */
+	if (sjme_error_is(error))
```
This can all be combined into a single statement.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -74,47 +533,233 @@ sjme_errorCode sjme_scritchpen_core_drawRect(
+	
+	/* The arcs cannot be larger then the rect's width/height*/
+	if(arcWidth >= w) { arcWidth = w-1; }
+	if(arcHeight >= h) { arcHeight = h-1; }
```
Code style, single line conditionals are on the following line with no extra `{}`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -74,47 +533,233 @@ sjme_errorCode sjme_scritchpen_core_drawRect(
 		return sjme_error_default(error);
+	
+	arcWidth = fabs(arcWidth);
+	arcHeight = fabs(arcHeight);
```
`abs()` and not `fabs()`, otherwise needless `float` conversion occurs.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -74,47 +533,233 @@ sjme_errorCode sjme_scritchpen_core_drawRect(
+	sjme_attrInNotNull sjme_scritchui_pencil g,
+	sjme_attrInValue sjme_jint xPoints[],
+	sjme_attrInPositive sjme_jint xOffset,
+	sjme_attrInValue sjme_jint yPoints[],
```
Must be pointers and not arrays here.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -74,47 +533,233 @@ sjme_errorCode sjme_scritchpen_core_drawRect(
+
+	if (g == NULL)
+		return SJME_ERROR_NULL_ARGUMENTS;
+		
```
`xOffset`, `yOffset`, and `nPoints` are not checked for bounds and overflow. They must be checked here.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -74,47 +533,233 @@ sjme_errorCode sjme_scritchpen_core_drawRect(
+		return sjme_error_default(error);
+	
+	/* Transform. TODO: This */
+	//sjme_scritchpen_coreUtil_applyTranslate(g, &x, &y);
```
Temporary copy through `sjme_alloca()` otherwise:

```
if (SJME_JNI_TRUE)
{
	sjme_todo("Impl?");
	return sjme_error_notImplemented(0);
}
```

All unimplemented code paths must end in failure.

AShiningRay replied:
That was pretty much cleared by today's conversation about how applyTranslate works. I can just uncomment that line.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -175,6 +818,94 @@ sjme_errorCode sjme_scritchpen_core_fillRect(
+		return sjme_error_default(error);
+	
+	arcWidth = fabs(arcWidth);
+	arcHeight = fabs(arcHeight);
```
`abs()` and not `fabs()` as otherwise needless `float` conversion occurs.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchPencilRasterShapes.c:
```diff
@@ -175,6 +818,94 @@ sjme_errorCode sjme_scritchpen_core_fillRect(
+	
+	/* The arcs cannot be larger than the rect's width/height*/
+	if(arcWidth >= w) { arcWidth = w-1; }
+	if(arcHeight >= h) { arcHeight = h-1; }
```
Code style, single line statements should be on the following line and there should be no extra `{}`.

XerTheSquirrel commented on nanocoat/lib/scritchui/scritchSerialize.c:
```diff
@@ -608,15 +618,15 @@ static sjme_thread_result sjme_attrThreadCall sjme_scritchui_serialDispatch(
 		as->drawRect.w,
 		as->drawRect.h));
 
-	SJME_SDP_CASE(drawTriangle,
-		SJME_SCRITCHUI_SERIAL_PEN_DRAW_TRIANGLE,
-		(as->drawTriangle.g,
-		as->drawTriangle.x1,
-		as->drawTriangle.y1,
-		as->drawTriangle.x2,
-		as->drawTriangle.y2,
-		as->drawTriangle.x3,
-		as->drawTriangle.y3));
```
Please bring back `drawTriangle()`.

XerTheSquirrel commented:
I have finished my preliminary review, nice work so far!

AShiningRay committed:
scritchPencilRaster: More improvements to Graphics methods.

This commit now implements drawTriangle and drawPolyline (same as
drawPolygon), as well as applying all requests made in the merge
review.

AShiningRay committed:
Drop long usage in fillPolygon.

It's not needed at all, and was just a relic of early implementation
testing.

AShiningRay committed:
Merge branch 'trunk' into moregraphicsmethods

XerTheSquirrel commented:
This has been merged as <https://squirreljme.cc/info/10e9707c40b64894>, thank you!
